inference-engine/src/inference_engine/ie_core.cpp

```diff
@@ -158,6 +158,148 @@ class Core::Impl : public ICore {
+        }
+
+    public:
+        using ConfigMap = std::map<std::string, std::string>;
+
+        CoreConfig() = default;
+
+        CoreConfig(const ICore * core,
+                   const std::string & deviceName,
+                   std::map<std::string, std::string> & config,
+                   const CoreConfig & globalConfig) {
+            // set global config settings
+            *this = globalConfig;
+
+            // parse local config
+            {
+                auto it = config.find(CONFIG_KEY(CACHE_DIR));
+                if (it != config.end()) {
+                    _config[it->first] = it->second;
+                    _modelCacheDir = it->second;
+                    _isModelCacheEnabled = true;
+
+                    // remove if plugin does support common IE Core key
+                    if (!checkPluginSupportsKey(core, it->first, deviceName))
+                        config.erase(it);
+                }
+            }
+        }
+
+        bool isModelCacheEnabled() const { return _isModelCacheEnabled; }
+        std::string getModelCacheDir() const { return _modelCacheDir; }
+    };
+
+    class CompiledBlobHeader final {
+        std::string _ie_version;
+
+    public:
+        CompiledBlobHeader() = default;
+
+        explicit CompiledBlobHeader(const std::string & ie_version) :
+            _ie_version(ie_version) {
+        }
+
+        const std::string & getIeVersion() const {
+            return _ie_version;
+        }
+
+        friend std::istream & operator >> (std::istream & stream, CompiledBlobHeader & header) {
+            std::string xmlStr;
+            std::getline(stream, xmlStr);
+
+            std::cerr << "Header " << std::endl;
+            std::cerr << xmlStr << std::endl;
+
+            pugi::xml_document document;
+            pugi::xml_parse_result res = document.load_string(xmlStr.c_str());
+
+            if (res.status != pugi::status_ok) {
+                THROW_IE_EXCEPTION_WITH_STATUS(NETWORK_NOT_READ) << "Error reading compiled blob header";
+            }
+
+            pugi::xml_node compiledBlobNode = document.document_element();
+            header._ie_version = XMLParseUtils::GetStrAttr(compiledBlobNode, "ie_version");
+
+            return stream;
+        }
+
+        friend std::ostream & operator << (std::ostream & stream, const CompiledBlobHeader & header) {
+            pugi::xml_document document;
+            auto compiledBlobNode = document.append_child("compiled_blob");
+            compiledBlobNode.append_attribute("ie_version").set_value(header._ie_version.c_str());
+
+            document.save(stream, nullptr, pugi::format_raw);
+            document.reset();
+            stream << std::endl;
+
+            return stream;
+        }
+    };
+
+    class CacheManager {
+    public:
+        using Ptr = std::shared_ptr<CacheManager>;
+        virtual ~CacheManager() = default;
+
+        virtual std::shared_ptr<std::ostream> writeCacheEntry(const std::string & id) = 0;
```

Comment:
can be unique_ptr, just don't use std::make_unique (since C++14), use instead either
return std::unique_ptr<std::ostream>(new std::ofstream(getBlobFile(blobHash), std::ios_base::binary));
or
stream.reset(new std::ifstream(blobFileName, std::ios_base::binary));

Reply:
This function API is point for discussion. I think better to have something like:
```
using StreamWriter = std::function<void(std::ostream&)>;
virtual void writeCacheEntry(const std::string & id, StreamWriter  writer) = 0;
```
And an user will use this function to fill the stream created and managed inside this function.

Because current API does not allow to encrypt stream ..

Reply:
Ok, I thought for encryption user will need to implement something like "OstreamAdapter" in this case